c/cardano_lock.c

```diff
@@ -1,18 +1,300 @@
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+
+  /* Clear lock field to zero, then digest the first witness
+   * lock_bytes_seg.ptr actually points to the memory in temp buffer
+   * */
+  memset((void *)lock_bytes_seg.ptr, 0, lock_bytes_seg.size);
+  blake2b_update(&blake2b_ctx, (char *)&witness_len, sizeof(uint64_t));
+  blake2b_update(&blake2b_ctx, temp, read_len);
+
+  // remaining of first witness
+  if (read_len < witness_len) {
+    ret = load_and_hash_witness(&blake2b_ctx, read_len, 0,
+                                CKB_SOURCE_GROUP_INPUT, false);
+    if (ret != CKB_SUCCESS) {
+      return ERROR_IDENTITY_SYSCALL;
+    }
+  }
+
+  // Digest same group witnesses
+  size_t i = 1;
+  while (1) {
+    ret =
+        load_and_hash_witness(&blake2b_ctx, 0, i, CKB_SOURCE_GROUP_INPUT, true);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_IDENTITY_SYSCALL;
+    }
+    i += 1;
+  }
+
+  // Digest witnesses that not covered by inputs
+  i = (size_t)ckb_calculate_inputs_len();
+  while (1) {
+    ret = load_and_hash_witness(&blake2b_ctx, 0, i, CKB_SOURCE_INPUT, true);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ERROR_IDENTITY_SYSCALL;
+    }
+    i += 1;
+  }
+
+  blake2b_final(&blake2b_ctx, msg, BLAKE2B_BLOCK_SIZE);
+
+  return 0;
+}
+
+#define CHECK(f, rc_code)                                       \
+  {                                                             \
+    bool flag = f;                                              \
+    if (!flag) {                                                \
+      ASSERT(false);                                            \
+      ckb_exit(rc_code);                                        \
+    }                                                           \
+  }
+
+int cardano_convert_copy(uint8_t *output, size_t *output_len,
+                         const uint8_t *payload, size_t payload_len,
+                         const uint8_t *external_aad, size_t external_aad_len) {
+  nanocbor_encoder_t enc;
+  nanocbor_encoder_init(&enc, output, *output_len);
+  nanocbor_fmt_array(&enc, 4);
+
+  const char *msg_sign_context = "Signature1";
+  nanocbor_put_tstr(&enc, msg_sign_context);
+  nanocbor_put_bstr(&enc, NULL, 0);
+  nanocbor_put_bstr(&enc, external_aad, external_aad_len);
+  nanocbor_put_bstr(&enc, payload, payload_len);
+
+  *output_len = nanocbor_encoded_len(&enc);
+  return CKB_SUCCESS;
+}
+
+int get_identity(uint8_t *identity) {
+  int err = CKB_SUCCESS;
+  unsigned char script[SCRIPT_SIZE];
+  uint64_t scritp_len = SCRIPT_SIZE;
+  err = ckb_load_script(script, &scritp_len, 0);
+  CHECK(err == CKB_SUCCESS, err);
+  CHECK(scritp_len <= SCRIPT_SIZE, ERROR_LOAD_SCRIPT);
+
+  mol_seg_t script_seg;
+  script_seg.ptr = (uint8_t *)script;
+  script_seg.size = scritp_len;
```

Comment:
These 3 lines can be combined: https://en.cppreference.com/w/c/language/struct_initialization